design-proposals/vhostuser-interface.md

```diff
@@ -0,0 +1,96 @@
+
+## Non Goals
+- The user space networking will require the host to a DPDK dataplane.
+- The kubevirtci creates kernel mode dataplane hosts for testing the VMs. It needs to be updated to support DPDK.
+
+## Definition of Users
+- Any user who has a (k8s) cluster with a dpdk dataplane can use the interface/feature.
+- The feature will be supported only if a valid CNI is being used.
+
+## User Stories
+- As a user/admin, I want to use user space networking feature if I have DPDK supported VM and the host has a DPDK dataplane.
+
+## Repos
+
+- kubevirt/kubevirt
+   - Introduction of vhostuser type interface.
+- kubevirt/kubevirtci
+   - Enable usage of dpdk on the setups created by kubevirtci.
+
+# Design
+- The design involves introducing vHostUser interface type and required parameters to support it. An EmptyDIR volume (shared-dir) and DownwardAPI (podinfo) volume are mounted on the virtlauncher pod to have the support for new interface.
+- These mounts will allow the virt-launcher pod to create a VM with an additional interface which can be reached using the vhostSocket mentioned in the DownwardAPI.
+
+- Creating the VMs will follow the same process with a few changes highlighted as below:
+    1. **Once the VM spec virt-controller will add two new volumes to the virt-launcher pod.**\
+        a. **EmptyDir volume named (shared-dir) is used to share the vhostuser socket file with the virt-launcher pod and dpdk dataplane.**\
+        b. **DownwardAPI volume named (podinfo) is used to share vhostuser socket file name with the virt-launcher via pod annotations.**
+    2. **The CNI will mount the shared-dir i.e., EmptyDir volume with the mount-path as "/var/lib/cni/usrcni/".**
+    3. **The CNI will update the virt-launcher pod annotations with vhostsocket-file name and details.**
+    4. **The virt-launcher reads the DownwardAPI volume and retrieves the vhostsocket-file name specified in pod annotations and uses it to create a unix socket in the VM, while launching the VM using libvirtd.**
+    4. **The virt-launcher is modified to skip establishing the networking between VM and the virt-launcher pod using bridge(Refer in Kubevirt Networking section in https://kubevirt.io/2018/KubeVirt-Network-Deep-Dive.html).**
+    5. **Instead of using the bridge through the launcher pod, the vHostUser interface of the VM will be directly connected to the DPDK datplane using the vhost socket**
+
+- With the above process a kubevirt VM with a vHostUser interface can be acheived. All the steps highlighted will be the changes made as part of the design.
+
+## API Examples
+
+Since, KubeVirt will always explicitly define the pod interface name for multus-cni. It will be computed from the VMI spec interface name, to allow multiple connections to the same multus provided network.
+
+The vHostUser Interface will be defined in the VM spec as  shown below:
+
+```yaml
+   devices:
+          interfaces:
+          - name: vhost-user-vn-blue
+            vhostuser: {}
```

Comment:
this implies adding yet another binding mechanism.

We've tried not to do that in the past (just a heads up ...) . 

We are actually trying to decouple this from KubeVirt, thus allowing end users to provide (and use ...) their bindings without moving the maintenance burden to KubeVirt.

Reply:
@maiqueb , Can you please give a few pointers on how this works. Thanks.

Reply:
Sure. 

1. update the [KubeVirt API](https://github.com/kubevirt/kubevirt/blob/24a67ecc90a44cc6109465bec109f216d1af2676/staging/src/kubevirt.io/api/core/v1/schema.go#L1237) with this new interface type.
2. AFAIU, nothing needs to be done for the phase#1 (virt-handler performs some privileged operations on behalf of the launcher). Everything you need should be done by userspace-cni, right ? 
3. the launcher domain manager **must** [generate the interface domXML](https://github.com/kubevirt/kubevirt/blob/24a67ecc90a44cc6109465bec109f216d1af2676/pkg/virt-launcher/virtwrap/converter/network.go#L90) snippet for your particular interface. More stuff done [here](https://github.com/kubevirt/kubevirt/blob/8c4fa60175c11257b955553a0328b44e86a044d4/pkg/network/domainspec/generators.go#L255) (example for macvtap).

Reply:
But as I've said in [this comment](https://github.com/kubevirt/community/pull/218#discussion_r1183788198), we're currently investigating how to allow users to extend the network binding options without requiring the core KubeVirt maintainers to maintain said bindings... We've recently discouraged users from posting new binding types ...

FYI .

@EdDev can provide more information; I just shipped around a few half baked ideas on how this could be done, but there's nothing concrete yet.

Reply:
@maiqueb, I believe I am not using a binding mechanism for the vhostuser interface, as I am not doing what exactly has been done for interfaces like bridge, masquerade, PASST.

I am using this PR https://github.com/kubevirt/kubevirt/pull/3208 as a reference and most of the changes are from it, I don't see any call following a binding mechanism.

Please let me know if you think this is still not clear.

Reply:
Hm, you may be right, seems we also don't need to generate the interface domXML snippet (on the launcher pod):
https://github.com/kubevirt/kubevirt/pull/3208/files#diff-030aa8b7ff9bf4a511d1d5641c1dec04205a944137672be3a691f0b1b4068753R93

Reply:
While pod networking configuration is not needed by virt-handler, the domain configuration is surely needed [1].

I am thinking that we can generalize these types of bindings. I.e noop bindings which do not touch the pod networking.
But we are still left with the logic of how to define the domain spec (libvirt domain config), including NUMA, hugepages, etc.

[1] https://github.com/kubevirt/kubevirt/pull/3208/files#diff-bc5b4c898f931f6ffeea71321777a71452dcb2e521946f77d28a72a885154a53

Reply:
@maiqueb, sorry for the confusion. I am actually using the binding interface method. Like I mentioned I made a few changes, and forgot I was using the interface binding method for vhostuser interface.

https://github.com/kubevirt/kubevirt/compare/release-0.59...Juniper:kubevirt:release-0.59

Reply:
@maiqueb , @EdDev 

Can you please confirm if we can proceed with the interface binding method, or do we have any work in progress for the same as mentioned in the  comment:
https://github.com/kubevirt/community/pull/218#discussion_r1183788198

Thanks